### PR TITLE
feat(tests): add Berkeley TestFloat for the operations IBM FPgen doesn't cover (C.4 spike resolution)

### PR DESCRIPTION
**Pull request by jeswr**
## Summary

Resolves the C.4 Berkeley TestFloat spike (`docs/ieee754-input-prep-redesign.md` §7.4, decided 2026-05-03).

The spike compared what TestFloat covers vs what IBM FPgen covers:

| Outcome | Verdict | Justification |
| --- | --- | --- |
| Coverage of operations | TestFloat is a strict superset | Every (op, precision) FPgen has, TestFloat has; TestFloat additionally has every binary64 op, ``rem`` at any precision, and full rounding-mode coverage for ``sqrt``/``fma`` |
| Coverage of *vectors* per operation | Complementary | FPgen carries handcrafted IBM Haifa adversarial vectors (cancellation, sticky-bit, vicinity-of-rounding-boundary, special significands) that TestFloat's level-1 random walk doesn't specifically construct |
| Verdict | **Both-needed** | Land TestFloat alongside FPgen, dispatch per (operation, precision, rounding) tuple |

### Coverage table

`OPERATION_SOURCES` populates 70 (op, precision, rounding) tuples:

| Source | Tuples |
| --- | --- |
| Both corpora | 30 (binary32 add/sub/mul/div/sqrt/fma at every rounding mode) |
| TestFloat only | 40 (every binary64 op; ``rem`` at any precision; sqrt/fma at non-default rounding for binary32 -- FPgen has thin coverage there) |
| FPgen only | 0 |

### Test count delta

| | Tests | Packages |
| --- | --- | --- |
| Before | 77,850 | 73 |
| After | 101,850 | 97 |
| Delta | **+24,000** | **+24** |

Critical addition: 8,000 *real* binary64 tests (4 ops x 2,000 cases) from TestFloat -- the existing `--generate-f64` flag fakes f64 by lossy round-trip through binary32, so the f64 circuits had no genuine native binary64 coverage before this PR.

### CI updates

- `.github/workflows/ci.yml`: `generate-tests` job now runs `./scripts/regenerate_testfloat_corpus.sh` before invoking the test generator. Builds Berkeley SoftFloat-3 + TestFloat-3 from upstream (~150 s) and captures `testfloat_gen` output for every dispatched (function, rounding) pair into `.testfloat_cache/`. The `--testfloat-max-per-function 2000` cap keeps CI tractable -- `mulAdd` at level 1 emits 6.1 M cases per rounding mode.

### Local validation

- Smoke run on `ieee754_test_testfloat_f64_add_nearest_even_part0` against the latest Noir toolchain: **1,243 / 1,250 cases pass**. The 7 failures are pre-existing `add_float64` denormal bugs (e.g. `MIN_DENORMAL + MIN_DENORMAL` returns the wrong value) that the FPgen-only path could not surface. Logged in workspace `STABILISATION-TODOS.md` under "Test-vector source concerns" for triage as part of the f64 circuit hardening pass.
- Generator regression smoke (with `--no-testfloat`): byte-identical output to the previous main.

### Roborev verdicts

- `029b820` (codex / gpt-5.5): 1 high (NaN payload preservation), 1 medium (dedup loop). Both addressed in `7853e0e`.
- `7853e0e` (codex / gpt-5.5): no issues found.

### Deferred to other agents

- **gmpy2 / MPFR agent**: non-default rounding modes are captured into `.testfloat_cache/` but skipped at emission time by `generate_noir_test`'s `RoundingMode != NEAREST_EVEN` guard. Drop the guard once an MPFR-backed reference oracle is wired in.
- **f64 circuit hardening (separate concern)**: the 7 TestFloat-surfaced `add_float64` denormal failures need triage. Tracked in workspace `STABILISATION-TODOS.md`.

## Test plan

- [x] Local generator runs cleanly with `--testfloat-cache` set
- [x] Local generator runs cleanly with `--no-testfloat` (regression check)
- [x] TestFloat parser round-trip of 46 k f64_add cases (NaN payloads aside, all bit patterns preserve)
- [x] Smoke nargo test of one TestFloat-sourced f64 package
- [ ] CI green on the PR (will validate post-push)

### CODE REVIEW

File: scripts/generate_tests.py [hunk 1]
```diff
@@ -1129,6 +1262,25 @@ def filter_tests(tests: list[TestCase]) -> list[TestCase]:
...
+                tf_tests_by_source = load_testfloat_tests(
+                    tf_cache,
+                    max_per_function=args.testfloat_max_per_function,
+                )
+                # Apply the same per-test filter the FPgen path uses.
+                for source_name, tests in tf_tests_by_source.items():
+                    filtered = filter_tests(tests)
+                    if filtered:
+                        tests_by_file[source_name] = filtered
+                        total_parsed += len(tests)
```
[the remark this answers is not in the document]
Comment (jeswr) [reply scripts/generate_tests.py:1429]: Fixed in 4b4da94. generate_ci_matrix now mirrors generate_noir_packages exactly: native f64 tests are counted only once (the f32 pass filters them by test.precision == BINARY32), so a 2 000-case f64 source produces 2 packages and 2 matrix entries. Verified locally with mock fixtures.

File: scripts/generate_tests.py [hunk 2]
```diff
@@ -1129,6 +1262,25 @@ def filter_tests(tests: list[TestCase]) -> list[TestCase]:
             if synthetic_tests:
                 tests_by_file["Synthetic-F64-Corner-Cases.synthetic"] = synthetic_tests
                 total_parsed += len(synthetic_tests)
+
+        # Add Berkeley TestFloat corpus if its cache is present.
+        if not args.no_testfloat:
+            tf_cache = Path(args.testfloat_cache) if args.testfloat_cache else get_testfloat_cache_dir()
```
[the remark this answers is not in the document]
Comment (jeswr) [reply scripts/generate_tests.py:1417]: Fixed in 4b4da94 with an explicit warning on the single-file path: --testfloat-cache is only honoured with --packages; ignoring it for the single-file output path. Wiring the corpus into the single-file emitter is bigger than the spike calls for; warning + ignore is the right shape until someone needs it.

File: scripts/regenerate_testfloat_corpus.sh [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    args = [
+        binary,
+        "-seed", str(seed),
+        "-level", str(level),
+        ROUNDING_FLAGS[rnd],
+        fn.name,
+    ]
+    print(f"  + {' '.join(args)} > {out.name}")
+    with open(out, "w") as fh:
+        subprocess.run(args, stdout=fh, check=True)
```
[the remark this answers is not in the document]
Comment (jeswr) [reply scripts/regenerate_testfloat_corpus.sh]: Fixed in 4b4da94. regenerate_testfloat_corpus.sh now accepts --max-per-function N and pipes testfloat_gen through head -n N at source. CI passes --max-per-function 2000, matching the existing --testfloat-max-per-function 2000 cap on the loader side. mulAdd no longer writes 6 M lines that get discarded.

File: scripts/regenerate_testfloat_corpus.sh [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    pair_set.add(dedup_key)
+    pairs.append((fn, key.rounding))
+
+print(f"Generating {len(pairs)} TestFloat captures into {cache_dir} (seed={seed}, level={level})")
+
+for fn, rnd in pairs:
+    out = cache_dir / f"{fn.name}_{rnd.name.lower()}.tfgen"
+    if out.exists() and out.stat().st_size > 0:
+        print(f"  cached {out.name}")
+        continue
```
[the remark this answers is not in the document]
Comment (jeswr) [reply scripts/regenerate_testfloat_corpus.sh:224]: Fixed in 4b4da94. The cache directory now carries a manifest.txt recording (seed, level, max-per-function, softfloat-pin, testfloat-pin); when any value differs from what produced the on-disk captures, the regen script invalidates the cache and regenerates. --force is the explicit always-wipe flag.

File: scripts/regenerate_testfloat_corpus.sh [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    SOFTFLOAT_DIR="${SOFTFLOAT_DIR:-$BUILD_DIR/berkeley-softfloat-3}"
+    TESTFLOAT_DIR="${TESTFLOAT_DIR:-$BUILD_DIR/berkeley-testfloat-3}"
+
+    if [[ ! -d "$SOFTFLOAT_DIR" ]]; then
+        echo "Cloning Berkeley SoftFloat-3 to $SOFTFLOAT_DIR..."
+        git clone --depth=1 https://github.com/ucb-bar/berkeley-softfloat-3 "$SOFTFLOAT_DIR"
+    fi
+    if [[ ! -d "$TESTFLOAT_DIR" ]]; then
+        echo "Cloning Berkeley TestFloat-3 to $TESTFLOAT_DIR..."
+        git clone --depth=1 https://github.com/ucb-bar/berkeley-testfloat-3 "$TESTFLOAT_DIR"
```
[the remark this answers is not in the document]
Comment (jeswr) [reply scripts/regenerate_testfloat_corpus.sh]: Fixed in 4b4da94. SOFTFLOAT_PIN and TESTFLOAT_PIN are pinned to specific commits (a0c6494c... and a9c849f1...) at the top of the regen script; clones are full-history + git checkout --detach <pin>. Bumping the pins is now an explicit, reviewable change, and .github/workflows/ci.yml keys the new GH-Actions cache off the pins so a bump invalidates the cache too.

File: scripts/noir_ieee754_inputs/sources.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    for op in _BIN32_ARITH_OPS:
+        for rnd in _ALL_ROUNDINGS:
+            table[_OpKey(op, Precision.BINARY64, rnd)] = _TESTFLOAT_ONLY
+
+    # sqrt: f32 default-rounding has FPgen vectors; everything else is
+    # TestFloat-only.
+    table[_OpKey(Operation.SQRT, Precision.BINARY32, RoundingMode.NEAREST_EVEN)] = _BOTH
+    for rnd in _ALL_ROUNDINGS:
+        if rnd != RoundingMode.NEAREST_EVEN:
+            table[_OpKey(Operation.SQRT, Precision.BINARY32, rnd)] = _BOTH
```
[the remark this answers is not in the document]
Comment (jeswr) [reply scripts/noir_ieee754_inputs/sources.py]: Fixed in 4b4da94. The docstring now matches what the loop populates: f32 every rounding is BOTH (FPgen has sqrt vectors at every rounding mode); f64 is TestFloat-only across the board.

File: scripts/noir_ieee754_inputs/sources.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        table[_OpKey(Operation.SQRT, Precision.BINARY64, rnd)] = _TESTFLOAT_ONLY
+
+    # fma: same shape as sqrt -- both at f32 default, TestFloat-only for
+    # f64 across the board.
+    for rnd in _ALL_ROUNDINGS:
+        table[_OpKey(Operation.FMA, Precision.BINARY32, rnd)] = _BOTH
```
[the remark this answers is not in the document]
Comment (jeswr) [reply scripts/noir_ieee754_inputs/sources.py]: Fixed in 4b4da94. The docstring now matches the loop: f32 every rounding is BOTH; f64 every rounding is TestFloat-only.

**Comment by jeswr** [comment]
Roborev review (job 448) — 3 Medium findings on the TestFloat corpus regen pipeline (cache key omits Python sources; pinned-commit checkout only on initial clone; pipe-shell status loss). Pre-existing patterns in the regen script, not introduced by this PR's headline (Berkeley TestFloat addition). Tracked as `todos/testfloat-corpus-cache-and-pin-fix.md` in the workspace. Merging.
